### PR TITLE
Implement contract deployment in Rust

### PR DESCRIPTION
**Pull request by nlordell**
This PR adds a new `contract` crate with a `deploy` script to deploy exchange contracts to a development network.

This an initial step towards #846 where we need an alternative to `scripts/setup_contracts.sh` which uses Truffle. The next steps in order to no longer require NodeJS are:
- Update `core` and `e2e` to use the new `contracts` crate.
- Update the Travis build file to use `cargo run -p contracts --bin deploy` instead of performing a Truffle migration.
- Use the `@gnosis.pm/dex-contracts` npm package for retrieving the contract artifact instead of building `dex-contracts` with Truffle.
- Update the README with new build instructions.

### Test Plan

Run the deploy script:
```
$ (cd contracts; cargo run -p contracts --bin deploy --features bin)
   Compiling contracts v0.1.0 (/var/home/nlordell/Developer/dex-services/contracts)
    Finished dev [unoptimized + debuginfo] target(s) in 12.14s
     Running `/var/home/nlordell/Developer/dex-services/target/debug/deploy`
[2020-08-05T12:43:07Z INFO  deploy] checking connection to local test node http://localhost:8545
[2020-08-05T12:43:07Z INFO  deploy] deploying library contracts
[2020-08-05T12:43:07Z INFO  deploy] deployed IdToAddressBiMap to 0x65dbc7c034b644401ad2a1f1ed8b284ae41e56bc
[2020-08-05T12:43:07Z INFO  deploy] deployed IterableAppendOnlySet to 0xc27e5a197f7f1c794bcd348f673d80687eff83a6
[2020-08-05T12:43:07Z INFO  deploy] deploying fee token contracts
[2020-08-05T12:43:07Z INFO  deploy] deployed TokenOWL to 0x8f534a8084d9306527ff15c919f7fa691c0ca856
[2020-08-05T12:43:07Z INFO  deploy] deployed TokenOWLProxy to 0x445d64d63f42401c8c7ff846d2a92edd2b31417b
[2020-08-05T12:43:07Z INFO  deploy] deploying exchange and viewer contracts
[2020-08-05T12:43:07Z INFO  deploy] deployed BatchExchange to 0x5355d6d088ddc1a99b29114976a6ca0321d71fa6
[2020-08-05T12:43:07Z INFO  deploy] deployed BatchExchangeViewer to 0x63f7170efe984d40d1a435a917b99c71ad645901
```

Then check that the `BatchExchange` contract address matches in the generated code:
```
$ for byte in $(echo '5355d6d088ddc1a99b29114976a6ca0321d71fa6' | grep -o ..); do printf "%d\n" "0x$byte"; done | xargs
83 85 214 208 136 221 193 169 155 41 17 73 118 166 202 3 33 215 31 166
$ cargo build && cat target/debug/build/contracts-*/out/BatchExchange.rs | grep -C 3 Address::from
   Compiling contracts v0.1.0 (/var/home/nlordell/Developer/dex-services/contracts)
    Finished dev [unoptimized + debuginfo] target(s) in 4.01s
            let artifact = Self::artifact();
            let network = match network_id {
                "5777" => self::ethcontract::common::truffle::Network {
                    address: self::ethcontract::Address::from([
                        83, 85, 214, 208, 136, 221, 193, 169, 155, 41, 17, 73, 118, 166, 202, 3,
                        33, 215, 31, 166,
                    ]),
```

### CODE REVIEW

**Review by bh2smith — COMMENTED** [review]
Looking good. Let me know when its ready. I suppose we will want to remove whatever reliance we currently have on truffle migrations previously used in this test env.

**Comment by nlordell** [comment]
> I suppose we will want to remove whatever reliance we currently have on truffle migrations previously used in this test env.

Exactly!

**Review by bh2smith — APPROVED** [review]

File: contracts/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    "log",
+]
+
+[dependencies]
+ethcontract = "0.7.2"
+serde = "1.0.114"
+
+# [bin-dependencies]
+anyhow = { version = "1.0.32", optional = true }
+env_logger = { version = "", optional = true }
```
Comment (e00E) [inline contracts/Cargo.toml:26]: intentional empty version string?

**Review by e00E — APPROVED** [review]
Nice to see this PR becoming ready!

File: contracts/src/bin/deploy.rs [hunk 2]
```diff
@@ -0,0 +1,120 @@
+//! Script to deploy Gnosis Protocol contracts to a local test network.
+//! Additionally writes the deployed addresses to the `target` directory so that
+//! they can be used by the build script.
```
Comment (e00E) [inline contracts/src/bin/deploy.rs:3]: Maybe there should be a short readme in the contracts folder that explains on a high level what we do in this crate and why and which paths are used.

File: contracts/src/paths.rs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+//! Module for common paths used for generated contract bindings.
+
+use std::path::{Path, PathBuf};
+
+/// Path to file containing address of a contract deployed to a test network.
+pub fn contract_address_file(name: &str) -> PathBuf {
```
Comment (fleupold) [inline contracts/src/paths.rs:6]: For rinkeby/mainnet, will we also need these `.addr` files or will we get the deployed addresses otherwise?

**Review by fleupold — APPROVED** [review]
<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia1.giphy.com%2Fmedia%2FtitioyQlWZmLK%2Fgiphy.gif&token=take_time" data-gitmeme-token="take_time"></a>

File: contracts/src/paths.rs (hunk 3 above)
Comment (nlordell) [reply contracts/src/paths.rs:6]: Rinkeby and mainnet addresses are included as part of the artifact JSON files (they get injected from the `networks.json` file). These are only needed for testnet deployments (the reason why we had to first run `./scripts/setup_contracts.sh` before building on Travis).

**Pull request merged**